includes/utils/class-amp-validation-utils.php

```diff
@@ -273,4 +501,185 @@ public static function display_error( $response ) {
 		echo '</div>';
 	}
 
+	/**
+	 * Whether to validate the front end response.
+	 *
+	 * @return boolean
+	 */
+	public static function should_validate_front_end() {
+		return ( self::has_cap() && ( isset( $_GET[ self::VALIDATION_QUERY_VAR ] ) ) ); // WPCS: CSRF ok.
+	}
+
+	/**
+	 * Adds the validation callback if front-end validation is needed.
+	 *
+	 * @param array $sanitizers The AMP sanitizers.
+	 * @return array $sanitizers The filtered AMP sanitizers.
+	 */
+	public static function add_validation_callback( $sanitizers ) {
+		if ( self::should_validate_front_end() ) {
+			foreach ( $sanitizers as $sanitizer => $args ) {
+				$args[ self::CALLBACK_KEY ] = __CLASS__ . '::track_removed';
+				$sanitizers[ $sanitizer ]   = $args;
+			}
+		}
+		return $sanitizers;
+	}
+
+	/**
+	 * Registers the post type to store the validation errors.
+	 *
+	 * @return void.
+	 */
+	public static function register_post_type() {
+		register_post_type(
+			self::POST_TYPE_SLUG,
+			array(
+				'labels'   => array(
+					'name' => _x( 'AMP Validation Errors', 'post type general name', 'amp' ),
+				),
+				'supports' => false,
+				'public'   => false,
+			)
+		);
+	}
+
+	/**
+	 * Stores the validation errors.
+	 *
+	 * After the preprocessors run, this gets the validation response if the query var is present.
+	 * It then stores the response in a custom post type.
+	 * If there's already an error post for the URL, but there's no error anymore, it deletes it.
+	 *
+	 * @return int|null $post_id The post ID of the custom post type used, or null.
+	 */
+	public static function store_validation_errors() {
+		if ( ! self::should_validate_front_end() ) {
+			return null;
+		}
```

Comment:
There should be a better way to limit who can make a request that stores validation errors. Right now, [should_validate_front_end()](https://github.com/Automattic/amp-wp/pull/971/files#diff-7d234989187d87a372e2aa6aa1299a71R510) checks for capability and the presence of the `validation` query var. 

But this needs to be able to make requests from the back-end. Like in [validate_home()](https://github.com/Automattic/amp-wp/pull/971/files#diff-7d234989187d87a372e2aa6aa1299a71R646), where it requests the front page on activating a plugin.

In that case, there won't be a logged-in user, as it's currently using [wp_remote_get()](https://github.com/Automattic/amp-wp/pull/971/files#diff-7d234989187d87a372e2aa6aa1299a71R647). Maybe there's a better way to make a request for a front-end page.

Reply:
One alternative is what was implemented in 4.9 for the scraping of plugin activation errors which is very similar to the validation checks we're doing in AMP. See:

* https://github.com/WordPress/wordpress-develop/blob/6e5a560be0c01e236b8d43da3ac61c50f34297d0/src/wp-admin/includes/file.php#L496-L505
* https://github.com/WordPress/wordpress-develop/blob/6e5a560be0c01e236b8d43da3ac61c50f34297d0/src/wp-includes/load.php#L1211-L1254

I don't know if we should go the route of storing the nonce in a transient.

On the other hand, you can see in the same example that `wp_remote_get()` can be passed the user's current cookies to make the request as an authenticated user:

https://github.com/WordPress/wordpress-develop/blob/6e5a560be0c01e236b8d43da3ac61c50f34297d0/src/wp-admin/includes/file.php#L525-L540

So that would allow for the capability check to continue to be used.

Reply:
I updated to pass cookies in fe0e6cc.